replacement_mode_modeling/04_FeatureClustering.ipynb

```diff
@@ -0,0 +1,1108 @@
+    "            'n_residents_with_license', 'income_category'\n",
+    "        ]:\n",
+    "            within_cluster_homogeneity[cix][col] = processed[col].std()\n",
+    "        else:\n",
+    "            within_cluster_homogeneity[cix][col] = entropy(processed[col])\n",
+    "\n",
+    "# Compute average homogeneity across other clusters.\n",
+    "for cix in within_cluster_homogeneity.keys():\n",
+    "    other_cluster_homogeneity[cix] = dict()\n",
+    "    other_clusters = set(within_cluster_homogeneity.keys()) - set([cix])\n",
+    "    for feature in within_cluster_homogeneity[cix].keys():\n",
+    "        homogeneity_in_others = [within_cluster_homogeneity[x][feature] for x in other_clusters]\n",
+    "        other_cluster_homogeneity[cix][feature] = np.mean(homogeneity_in_others)\n",
+    "\n",
+    "        \n",
+    "# Compute contrastive homogeneity\n",
+    "# CH = homogeneity within cluster / average homogeneity across other clusters\n",
+    "for cix in within_cluster_homogeneity.keys():\n",
+    "    ch_scores = list()\n",
+    "    print(f\"For cluster {cix}:\")\n",
+    "    for feature in within_cluster_homogeneity[cix].keys():\n",
+    "        feature_ch = within_cluster_homogeneity[cix][feature]/(other_cluster_homogeneity[cix][feature] + 1e-6)\n",
+    "        ch_scores.append((feature, feature_ch))\n",
+    "    \n",
+    "    ch_df = pd.DataFrame(ch_scores, columns=['feature', 'ch']).sort_values(by=['ch']).head(4)\n",
+    "    \n",
+    "    # Display actual values.\n",
+    "    users = target_df[labels == cix].index\n",
+    "    data = demographics.loc[demographics.index.isin(users), :].reset_index(drop=True)\n",
+    "    processed = preprocess_demo_data(data)\n",
+    "    \n",
+    "    display(ch_df)\n",
+    "    print()\n",
+    "    filtered = processed.loc[:, processed.columns.isin(ch_df.feature)][ch_df.feature]\n",
+    "    filtered_features = ch_df.feature.tolist()\n",
+    "    \n",
+    "    fig, ax = plt.subplots(nrows=2, ncols=2, figsize=(12, 10))\n",
+    "    for i, a in enumerate(ax.flatten()):\n",
+    "        sns.histplot(filtered[filtered_features[i]], ax=a, stat=\"percent\")\n",
+    "    plt.tight_layout()\n",
+    "    plt.savefig(f\"{CURRENT_DB}_{cix}_Demographic_consistency.png\", dpi=300)\n",
+    "    plt.show()\n",
+    "    print()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "580bbd86",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from scipy.stats import iqr\n",
+    "\n",
+    "def get_trip_summary_df(users, df):\n",
+    "    '''\n",
+    "    1. df = a huge dataframe of user-trips. Each row is a trip.\n",
+    "    2. every trip is divided into sections: [walk, transit, walk]\n",
+    "    3. Each section has a corresponding distance and duration: [m1, m2, m3], [t1, t2, t3], [d1, d2, d3]\n",
+    "    4. What we are doing is only considering the mode, distance, and duration of the section with the largest distance\n",
```

Comment:
Does this mean the summary ends up being just one segment of the trip? Or is the total of all segments with the most represented mode?

Reply:
For every user, the trip summary is a combination of the aggregation performed over the 'argmax' section information and the entire trip-level distance and duration aggregates.